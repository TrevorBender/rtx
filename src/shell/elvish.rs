use std::path::Path;

use indoc::formatdoc;

use crate::shell::Shell;

#[derive(Default)]
pub struct Elvish {}

impl Shell for Elvish {
    fn activate(&self, exe: &Path, flags: String) -> String {
        let exe = exe.to_string_lossy();
        let mut out = String::new();
        out.push_str(&formatdoc! {r#"
            set-env RTX_SHELL elvish

            edit:add-var rtx~ {{|@args|
              var command
              if (> (count $args) 0) {{
                set command = $args[0]
              }}
              if (or (eq $command "deactivate") (eq $command "shell")) {{
                eval ({exe} $@args | slurp)
              }} else {{
                {exe} $@args
              }}
            }}

            fn rtx_hook {{||
              eval ({exe} hook-env{flags} -s elvish | slurp)
            }}
            edit:add-var _rtx_hook $rtx_hook~

            set edit:before-readline = [ $@edit:before-readline $rtx_hook~ ]
            rtx_hook
            "#});

        out
    }

    fn deactivate(&self) -> String {
        formatdoc! {r#"
            unset-env MISE_SHELL
            set edit:before-readline = [(each {{|f|
              if (not-eq $f $_rtx_hook) {{
                put $f
              }}
            }} $edit:before-readline)]
            del rtx~
        "#}
    }

    fn set_env(&self, k: &str, v: &str) -> String {
        let k = shell_escape::unix::escape(k.into());
        let v = shell_escape::unix::escape(v.into());
        let v = v.replace("\\n", "\n");
        format!("set-env {k} {v}\n")
    }

    fn prepend_env(&self, k: &str, v: &str) -> String {
        format!("set-env {k} {v}\n")
    }

    fn unset_env(&self, k: &str) -> String {
        format!("unset-env {k}\n", k = shell_escape::unix::escape(k.into()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hook_init() {
        let elvish = Elvish::default();
        let exe = Path::new("/some/dir/rtx");
        assert_snapshot!(elvish.activate(exe, " --status".into()));
    }
}
