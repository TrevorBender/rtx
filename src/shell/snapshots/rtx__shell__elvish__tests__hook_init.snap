---
source: src/shell/elvish.rs
expression: "elvish.activate(exe, true)"
---
set paths = [ /some/dir $@paths ]
set-env RTX_SHELL elvish

fn _rtx_hook {
  eval (/some/dir/rtx hook-env --status -s elvish | slurp)
}

set edit:before-readline = [ $@edit:before-readline $_rtx_hook~ ]
_rtx_hook

